### PR TITLE
Add snapshot shard size based test in DiskThresholdDeciderTests

### PR DESCRIPTION
**Pull request by tlrx**
This PR adds a test in `DiskThresholdDeciderTests` that verifies the allocation of a snapshot recovery source based shard in the situation where the snapshot shard size was successfully provided by the SnapshotInfoService introduced in #61906 and when the service failed to provide the size.

Relates https://github.com/elastic/elasticsearch/pull/61906#pullrequestreview-502046246.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Sorry I missed this one Tanguy, will take a look today.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java [hunk 1]
```diff
@@ -1130,6 +1144,110 @@ public void testWatermarksEnabledForSingleDataNode() {
...
+                makeDecider(diskSettings))));
+
+        final Snapshot snapshot = new Snapshot("_repository", new SnapshotId("_snapshot_uuid", "_snapshot_name"));
+        final IndexId indexId = new IndexId("_indexid_uuid", "_indexid_name");
+        final ShardId shardId = new ShardId(new Index("test", IndexMetadata.INDEX_UUID_NA_VALUE), 0);
+
+        final AtomicReference<SnapshotShardSizeInfo> snapshotShardSizeInfoRef = new AtomicReference<>(SnapshotShardSizeInfo.EMPTY);
+        final SnapshotsInfoService snapshotsInfoService = snapshotShardSizeInfoRef::get;
+
+        final AllocationService strategy = new AllocationService(deciders, new TestGatewayAllocator(),
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java:1178]: NIT: I think I would have had an easier time following this test if we had moved this to where it's actually first used after the assertions on the CS we construct below to like:

```java
        final AtomicReference<SnapshotShardSizeInfo> snapshotShardSizeInfoRef = new AtomicReference<>(SnapshotShardSizeInfo.EMPTY);
        final AllocationService strategy = new AllocationService(deciders, new TestGatewayAllocator(),
                new BalancedShardsAllocator(Settings.EMPTY), clusterInfoService, snapshotShardSizeInfoRef::get);

        // reroute triggers snapshot shard size fetching
        clusterState = strategy.reroute(clusterState, "reroute");
        logShardStates(clusterState);
```

**Review by original-brownbear — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java [hunk 2]
```diff
@@ -1130,6 +1144,110 @@ public void testWatermarksEnabledForSingleDataNode() {
...
+
+        final AllocationDeciders deciders = new AllocationDeciders(
+            new HashSet<>(Arrays.asList(
+                new RestoreInProgressAllocationDecider(),
+                new SameShardAllocationDecider(
+                    Settings.EMPTY, new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS)
+                ),
+                makeDecider(diskSettings))));
+
+        final Snapshot snapshot = new Snapshot("_repository", new SnapshotId("_snapshot_uuid", "_snapshot_name"));
```
[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java:1171]: Good catch

[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java:1171]: Let's correct the IndexId as well :facepalm:

**Comment by tlrx** [comment]
Thanks Armin!

**Pull request merged**